### DISCUSSION TITLE
combining docker-slim and buildx

### DISCUSSION THREAD

**Discussion by rotty3000** [slimtoolkit/slim#435] [Q&A]
Is it possible to combine docker-slim with buildx?

We're working on a developer experience where we package many tools into a docker image which is then dispatched to from a native cli. In order to support Mac M1 we need to build that tools image with buildx. However it would be really nice if we could then also benefit from docker-slim to reduce download sizes.

**Comment by kcq** [comment]
Just to double check... You want DockerSlim to use buildx when it generated the minified image instead of using the standard docker build API, right?

**Comment by rotty3000** [comment]
That's a great question. Perhaps my problem is in understanding the workflow.

Today we use `docker buildx` which means the outputted images are in our registry (not local).

What would be your recommended workflow from there? Remember that we have multi-arch images and I'm guessing docker-slim would have to operate on each one separately.

**Comment by kcq** [comment]
@rotty3000 There are a few things here that will be good to go over... 

First, making sure the output image is in the registry can be done in a couple of ways. The easiest way to do it (which is how I normally do it) is to add a `docker push` line after `slim build` to put the minified image in my registry. This gets slightly more involved when you deal with multi-architecture images with fat manifests (i'll add more about that later here). Another option is to add an enhancement to the `build` command and a command flag that would allow `slim` to push the generated output image to a registry. The reason this wasn't there from day one is because now the tool needs to deal with credentials and you need to make sure you point `slim` to the right credentials, which is doable, but it's extra work and, in some cases, more friction than simply relying on `docker` or another tool to push the local image.

Next, i'd like to mention the relatively new `--image-build-engine` and `--image-build-arch` build command flags. The `--image-build-engine` flag allows you to pick different output image builders. `docker` is the default builder, but you can also pick `internal`, which is a standalone builder with no external dependencies on docker (nice in many cases, a feature towards not requiring Docker as a dependency at all). If you pick `none` then `slim` won't build the output image for you at all, but you'll still have the artifacts to build an output image yourself any way you want (in the artifacts directory you'll have `files.tar` and `Dockerfile` for that). One of the builders here will be `buildx`, but it hasn't been fully implemented yet (though there's an outstanding PR that's related to it).

Finally, dealing with the multi-architecture images... The most straight forward way to deal with those is to minify the per-architecture image separately. Push each per-architecture image to your registry. Then create a multi-architecture manifest pointing to those minified per-architecture images. There's a number of ways to do it including the `docker manifest create` command (e.g., https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/ ). Future versions of `slim` will include this capability internally (it's definitely nice to have, just not enough cycles at the moment to prioritize it, but very open to collaborating on this).
